pkg/chatbot/twitch_chatbot.go

```diff
@@ -0,0 +1,116 @@
+package chatbot
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gempir/go-twitch-irc/v2"
+	"github.com/nicklaw5/helix/v2"
+	"github.com/saxypandabear/twitchsongrequests/internal/constants"
+	"github.com/saxypandabear/twitchsongrequests/pkg/db"
+	"github.com/zmb3/spotify/v2"
+	"go.uber.org/zap"
+	"golang.org/x/oauth2"
+)
+
+func SendChatMessage(broadcaster string, userStore db.UserStore, event *helix.EventSubChannelPointsCustomRewardRedemptionEvent) error {
+	u, err := userStore.GetUser(event.BroadcasterUserID)
+	if err != nil {
+		return err
+	}
+
+	chatClient := twitch.NewClient(broadcaster, "oauth:"+u.TwitchAccessToken)
+	chatClient.Join(broadcaster)
+
+	chatClient.OnConnect(func() {
+		spotifyToken := &oauth2.Token{
+			AccessToken:  u.SpotifyAccessToken,
+			RefreshToken: u.SpotifyRefreshToken,
+			Expiry:       *u.SpotifyExpiry,
+		}
+
+		spotifyConfig := &oauth2.Config{
+			ClientID:     os.Getenv(constants.SpotifyClientIDKey),
+			ClientSecret: os.Getenv(constants.SpotifyClientSecretKey),
+			Endpoint: oauth2.Endpoint{
+				AuthURL:  "https://accounts.spotify.com/authorize",
+				TokenURL: "https://accounts.spotify.com/api/token",
+			},
+		}
+
+		httpClient := spotifyConfig.Client(context.Background(), spotifyToken)
+		spotifyClient := spotify.New(httpClient)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+
+		track, trackErr := getSpotifyTrack(ctx, spotifyClient, event.UserInput)
+		if trackErr != nil {
+			zap.L().Error("Error in Spotify search or retrieving the track", zap.Error(trackErr))
+			chatClient.Say(broadcaster, "There was an error searching for the song.")
+			return
+		}
+
+		if track != nil {
+			sendTrackMessage(chatClient, broadcaster, track, event.UserName)
+		} else {
+			chatClient.Say(broadcaster, "No matching song was found.")
+		}
+	})
+
+	err = chatClient.Connect()
+	if err != nil {
+		zap.L().Error("Error connecting to Twitch chat",
+			zap.String("id", event.BroadcasterUserID),
+			zap.String("broadcaster", broadcaster),
+			zap.Error(err))
+		return err
+	}
+
+	return nil
```

Comment:
I'm a little confused. So we create the chat client, define a callback so when it connects, it immediately sends a message, and then return. Do we need to also close the client? I'm worried about leaking resources by creating a bunch of Twitch clients.

Reply:
I understand your concerns about the potential leakage of resources from the creation of multiple Twitch clients. You are right that it is important to manage resources properly.

Here are my solutions:
- We create a channel messageSent to signal when the message has been sent.
- In the OnConnect function, we close the channel after the message has been sent.
- After chatClient.Connect() we wait for the signal from the messageSent channel.
- Finally, we disconnect the client properly.

I actually dont know how to show the changes properly, so i just send the code snippet into the Suggestion.

```suggestion
	func SendChatMessage(broadcaster string, userStore db.UserStore, event *helix.EventSubChannelPointsCustomRewardRedemptionEvent) error {
	u, err := userStore.GetUser(event.BroadcasterUserID)
	if err != nil {
		return err
	}

	chatClient := twitch.NewClient(broadcaster, "oauth:"+u.TwitchAccessToken)
	chatClient.Join(broadcaster)

	messageSent := make(chan struct{})

	chatClient.OnConnect(func() {
		spotifyToken := &oauth2.Token{
			AccessToken:  u.SpotifyAccessToken,
			RefreshToken: u.SpotifyRefreshToken,
			Expiry:       *u.SpotifyExpiry,
		}

		spotifyConfig := &oauth2.Config{
			ClientID:     os.Getenv(constants.SpotifyClientIDKey),
			ClientSecret: os.Getenv(constants.SpotifyClientSecretKey),
			Endpoint: oauth2.Endpoint{
				AuthURL:  "https://accounts.spotify.com/authorize",
				TokenURL: "https://accounts.spotify.com/api/token",
			},
		}

		httpClient := spotifyConfig.Client(context.Background(), spotifyToken)
		spotifyClient := spotify.New(httpClient)

		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()

		track, trackErr := getSpotifyTrack(ctx, spotifyClient, event.UserInput)
		if trackErr != nil {
			zap.L().Error("Error in Spotify search or retrieving the track", zap.Error(trackErr))
			chatClient.Say(broadcaster, "There was an error searching for the song.")
			return
		}

		if track != nil {
			sendTrackMessage(chatClient, broadcaster, track, event.UserName)
		} else {
			chatClient.Say(broadcaster, "No matching song was found.")
		}

		close(messageSent)
	})

	err = chatClient.Connect()
	if err != nil {
		zap.L().Error("Error connecting to Twitch chat",
			zap.String("id", event.BroadcasterUserID),
			zap.String("broadcaster", broadcaster),
			zap.Error(err))
		return err
	}

	<-messageSent

	if err := chatClient.Disconnect(); err != nil {
		zap.L().Error("Error disconnecting from Twitch chat",
			zap.String("broadcaster", broadcaster),
			zap.Error(err))
		return err
	}

	return nil
}
```

Reply:
That makes sense to me!